### PR TITLE
Adding bist afu support to the fpgabist tool

### PR DESCRIPTION
**Pull request by buie1**
The fpgabist tool will now be able to use the bist_afu as one of the modes.  

There will be additional development on the application to remove warnings.

### CODE REVIEW

File: tools/fpgabist/bist/bist_afu.c [hunk 1]
```diff
@@ -256,18 +256,17 @@ int main(int argc, char *argv[])
...
 	printf("Running BIST Test\n");
         uint64_t data = 0;
-        uint64_t fpga_res = 0;
         double count = 0;
         unsigned int bist_mask = ENABLE_DDRA_BIST;
         unsigned int ddra_bist_result = 0;
 
 	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
 
-        fpga_res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_CTRL_ADDR,&data);
```
Comment (southerngs) [inline tools/fpgabist/bist/bist_afu.c]: why don't you want to check the return code here?

File: tools/fpgabist/bist/bist_afu.c [hunk 2]
```diff
@@ -278,10 +277,10 @@ int main(int argc, char *argv[])
 	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
         while ((CHECK_BIT(data,9) != 0x200) && (CHECK_BIT(data,8) != 0x100) && (CHECK_BIT(data,7) != 0x80) && 
             (CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
-          fpga_res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
```
Comment (southerngs) [inline tools/fpgabist/bist/bist_afu.c]: same comment about return code check

File: tools/fpgabist/bist/bist_afu.c [hunk 3]
```diff
@@ -256,18 +256,17 @@ int main(int argc, char *argv[])
...
         unsigned int bist_mask = ENABLE_DDRA_BIST;
         unsigned int ddra_bist_result = 0;
 
 	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
 
-        fpga_res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_CTRL_ADDR,&data);
+        fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_CTRL_ADDR,&data);
 	while((CHECK_BIT(data,27) != 0x08000000)){
-	  printf("Enable Test: reading result #%f: %04x\n", count, data);
+	  printf("Enable Test: reading result #%f: %04x\n", count, (unsigned int)data);
```
Comment (southerngs) [inline tools/fpgabist/bist/bist_afu.c]: you should probably use `%lx` in printf formatting rather than using a cast here (in general avoid unnecessary casting and don't use casting to silence warnings except in rare cases).

File: tools/fpgabist/bist/bist_afu.c [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+                 &data); 
+          if (count >= MAX_COUNT){
+		fprintf(stderr, "DDR Bank B BIST Timed Out.\n");
+                break;
+          }
+          count++;
+	  usleep(100);
+        }
+
+        if (ddra_bist_result == DDRA_BIST_PASS) {
```
Comment (dasm) [inline tools/fpgabist/bist/bist_afu.c]: Why are you doing this? Instead of assigning new value `ddra_bist_result` just move this logic to lines L290-298 and you'll have the same thing.

File: tools/fpgabist/bist/bist_afu.c [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+/************* Begin BIST *************/
+        /* Perform BIST Check */
+	printf("Running BIST Test\n");
+        uint64_t data = 0;
+        double count = 0;
+        unsigned int bist_mask = ENABLE_DDRA_BIST;
+        unsigned int ddra_bist_result = 0;
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
```
Comment (dasm) [inline tools/fpgabist/bist/bist_afu.c]: Why don't you create separate functions performing check?

File: tools/fpgabist/bist/bist_afu.c [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+#define DDRB_BIST_PASS 0x1000
+#define DDRB_BIST_FAIL 0x800
+#define DDRB_BIST_TIMEOUT 0x400
+#define DDRB_BIST_FATAL_ERROR 0x20
+/**************** BIST #defines *****************/
+
+
+/* NLB0 AFU_ID */
+//#define NLB0_AFUID "D8424DC4-A4A3-C413-F89E-433683F9040B"
+#define NLB0_AFUID "9caef53d-2fcf-43ea-84b9-aad98993fe41"
```
[the remark this answers is not in the document]
Comment (southerngs) [reply tools/fpgabist/bist/bist_afu.c]: @dasm how do we get the AFU ID then?  This is what the hello_fpga sample does, but maybe there's a better way now.

File: tools/fpgabist/bist/bist_afu.c (hunk 5 above)
Comment (buie1) [reply tools/fpgabist/bist/bist_afu.c]: Right now with the GOTO's abstaction of C code is increasingly complicated.  There will need to be some future effort to address these concerns in future applications.

File: tools/fpgabist/README [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+- fpgainfo
+
+Accepted gbs files:
+- bist_afu.gbs
+- dma_afu.gbs
+- nlb_mode_3.gbs
+
+Usage:
+fpgabist -b <bus_num> <path to gbs file>
+eg. fpgabist -b 6 ../../../dcp_1.1/samples/nlb_mode_3/bin/nlb_mode_3.gbs
```
Comment (r-rojo) [inline tools/fpgabist/README]: is the bus usually expressed in hex or it doesn't matter?

File: tools/fpgabist/bist/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+## CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+
+include_directories(${OPAE_INCLUDE_DIR}
+                    ${OPAE_SDK_SOURCE}/libopae/src )
+file(GLOB CSources *.c)
```
Comment (r-rojo) [inline tools/fpgabist/bist/CMakeLists.txt]: Can we explicitly call out the source files rather than relying on the output of GLOB?

File: tools/fpgabist/bist/README [hunk 9]
```diff
@@ -0,0 +1,7 @@
+The Arria® 10 DCP BIST is a mechanism designed to allow 
+the DCP system to troubleshoot itself and find components
```
Comment (r-rojo) [inline tools/fpgabist/bist/README]: Is this limited to only DCP systems? Would it not work for Xeon+Integrated FPGA?

File: tools/fpgabist/bist/bist_afu.c [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+
+	count = 0;
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,9) != 0x200) && (CHECK_BIT(data,8) != 0x100) && (CHECK_BIT(data,7) != 0x80) && 
+            (CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
+          res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
+                 &data); 
+	  ON_ERR_GOTO(res, out_free_output, "Reading DDR_BIST_STATUS");
+          if (count >= MAX_COUNT){
+		fprintf(stderr, "DDR Bank A BIST Timed Out.\n");
```
Comment (r-rojo) [inline tools/fpgabist/bist/bist_afu.c]: Alignment seems off

File: tools/fpgabist/bist/bist_afu.c [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+          }
+	  count++;
+	}
+	printf("BIST is enabled.  Reading status register\n");
+
+	count = 0;
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,9) != 0x200) && (CHECK_BIT(data,8) != 0x100) && (CHECK_BIT(data,7) != 0x80) && 
+            (CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
+          res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
```
Comment (r-rojo) [inline tools/fpgabist/bist/bist_afu.c]: Can this `while` condition be simplified? Maybe in a follow up PR?

File: tools/fpgabist/bist/bist_afu.c [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+		printf("DDR Bank A BIST Test timed out.\n");
+        } else {
+		printf("DDR Bank A Test encountered a fatal error and cannot continue.\n");
+        }
+
+        bist_mask = ENABLE_DDRB_BIST;
+        count = 0;
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
```
Comment (r-rojo) [inline tools/fpgabist/bist/bist_afu.c]: Same thing here. Can this `while` condition be simplified? Perhaps a bitfield struct?

File: tools/fpgabist/bist/bist_afu.c [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+          res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
+                 &data);
+	  ON_ERR_GOTO(res, out_free_output, "Reading DDR_BIST_STATUS_ADDR");
+          if (count >= MAX_COUNT){
+		fprintf(stderr, "DDR Bank B BIST Timed Out.\n");
+                break;
```
Comment (r-rojo) [inline tools/fpgabist/bist/bist_afu.c]: Indentation seems off

File: tools/fpgabist/bist/bist_afu.c [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+          }
+          count++;
+	  usleep(100);
```
Comment (r-rojo) [inline tools/fpgabist/bist/bist_afu.c]: Indentation seems off

File: tools/fpgabist/bist_common.py [hunk 15]
```diff
@@ -31,7 +31,7 @@
 import sys
 
 # TODO: Use AFU IDs vs. names of AFUs
```
Comment (r-rojo) [inline tools/fpgabist/bist_common.py]: 👍

File: tools/fpgabist/README (hunk 7 above)
Comment (buie1) [reply tools/fpgabist/README]: bus can be expressed in hex such as 0x06 or 06

File: tools/fpgabist/bist/CMakeLists.txt (hunk 8 above)
Comment (buie1) [reply tools/fpgabist/bist/CMakeLists.txt]: this can be changed since there's only one so far.

File: tools/fpgabist/bist/bist_afu.c (hunk 11 above)
Comment (buie1) [reply tools/fpgabist/bist/bist_afu.c]: future changes like this are going to be made in future PRs

File: tools/fpgabist/bist/bist_afu.c (hunk 12 above)
Comment (buie1) [reply tools/fpgabist/bist/bist_afu.c]: ditto

File: tools/fpgabist/bist/bist_afu.c (hunk 11 above)
Comment (r-rojo) [reply tools/fpgabist/bist/bist_afu.c]: Right, that's why I say "in a follow up PR". How are you tracking these TODO things?

**Comment by r-rojo** [comment]
If we were to add tools/fpgabist to the list of directories that test-codingstyle.sh looks at, the checkpatch.pl script fails (with almost 400 errors). Do we want to add it to test-codingstyle.sh and fix the errors?

**Comment by southerngs** [comment]
@r-rojo do you know how many new errors are introduced by this PR?  Eventually we should fix all of them, but probably not today.  We want to get this PR in today if possible.

**Comment by r-rojo** [comment]
doing a "test-codingstyle.sh | wc -l" results in a line count of 387. That is a lot to do in a time crunch. I'm okay with deferring this fix for another pull request as long as this is tracked somehow and it's addressed soon.

File: tools/fpgabist/bist/bist_afu.c [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <uuid/uuid.h>
+#include <opae/fpga.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+int usleep(unsigned);
```
Comment (southerngs) [inline tools/fpgabist/bist/bist_afu.c]: why is this here?  You should already have the declaration from unistd.h

**Comment by buie1** [comment]
@r-rojo Going to track the coding style fixes in future PR
https://github.com/OPAE/opae-sdk/pull/246

**Comment by southerngs** [comment]
@r-rojo do you know if someone can sign-off for ci?

**Comment by southerngs** [comment]
also is one of the maintainers supposed to do the actual merge?

**Comment by r-rojo** [comment]
@rpSTTD and @nakulkorde  are in @OPAE/ci 
They seem to be offline now...

**Comment by r-rojo** [comment]
@southerngs , yeah. I think one of the owners is supposed to merge.
